### PR TITLE
Fix module stateful submodule

### PR DESCRIPTION
**Pull request by RickSanchezStoic**
fix_module_stateful_submodule

### CODE REVIEW

File: ivy/stateful/module.py [hunk 1]
```diff
@@ -152,6 +181,14 @@ def _top_mod_fn(self, depth=None):
 
     # noinspection PyProtectedMember
     def track_submod_rets(self):
+        """
+        Tracks the returns of the submodules if track_submod_returns 
+        argument is set to True during call
```
Comment (MrZilinXiao) [inline ivy/stateful/module.py:186]: don't think this func works as the comment tells

**Comment by RickSanchezStoic** [comment]
@djl11 Currently the following functions don't have a docstring `mod_with_top, _mod_key_chain
_add, _submod_enter,
_extract_v`. Rest all have been added by me Zilin and Jin from what we could figure out. Kindly have a look and let me know if more changes are needed. The tests are passing except for the module training for jax backend, it takes a little longer to train so will have to increase the hypothesis' time limit to about 5000 ms.

File: ivy_tests/test_ivy/test_stateful/test_modules.py [hunk 2]
```diff
@@ -21,23 +23,31 @@ def __init__(
...
+        x = ivy.expand_dims(x, axis=0)
         x = ivy.tanh(self._linear0(x))
         x = ivy.tanh(self._linear1(x))
         return ivy.tanh(self._linear2(x))[0]
 
 
 # module training
-@pytest.mark.parametrize("bs_ic_oc", [([1, 2], 4, 5)])
-def test_module_training(bs_ic_oc, device, compile_graph):
+# @pytest.mark.parametrize("bs_ic_oc", [([1, 2], 4, 5)])
```
Comment (djl11) [inline ivy_tests/test_ivy/test_stateful/test_modules.py:36]: this comments can be removed

File: ivy_tests/test_ivy/test_stateful/test_modules.py [hunk 3]
```diff
@@ -82,23 +92,31 @@ def __init__(self, in_size, out_size, device=None, hidden_size=64):
...
+        x = ivy.expand_dims(x, axis=0)
         x = ivy.tanh(self._layers[0](x))
         x = ivy.tanh(self._layers[1](x))
         return ivy.tanh(self._layers[2](x))[0]
 
 
 # module with list training
-@pytest.mark.parametrize("bs_ic_oc", [([1, 2], 4, 5)])
-def test_module_w_list_training(bs_ic_oc, device, compile_graph):
+# @pytest.mark.parametrize("bs_ic_oc", [([1, 2], 4, 5)])
```
Comment (djl11) [inline ivy_tests/test_ivy/test_stateful/test_modules.py:105]: this comments can be removed

File: ivy_tests/test_ivy/test_stateful/test_modules.py [hunk 4]
```diff
@@ -138,13 +156,26 @@ def loss_fn(v_):
 
 
 # module with partial v
-@pytest.mark.parametrize("bs_ic_oc", [([1, 2], 4, 5)])
-def test_module_w_partial_v(bs_ic_oc, device, compile_graph):
+# @pytest.mark.parametrize("bs_ic_oc", [([1, 2], 4, 5)])
```
Comment (djl11) [inline ivy_tests/test_ivy/test_stateful/test_modules.py:159]: this comments can be removed

File: ivy_tests/test_ivy/test_stateful/test_modules.py [hunk 5]
```diff
@@ -202,20 +233,28 @@ def test_module_w_partial_v(bs_ic_oc, device, compile_graph):
...
+        ivy.Module.__init__(self, device=device)
 
     def _forward(self, x):
         return x
 
 
 # module with none attribute
-@pytest.mark.parametrize("bs_ic_oc", [([1, 2], 4, 5)])
-def test_module_w_none_attribute(bs_ic_oc, device, compile_graph):
+# @pytest.mark.parametrize("bs_ic_oc", [([1, 2], 4, 5)])
```
Comment (djl11) [inline ivy_tests/test_ivy/test_stateful/test_modules.py:243]: this comments can be removed

File: ivy_tests/test_ivy/test_stateful/test_modules.py [hunk 6]
```diff
@@ -246,14 +285,22 @@ def _forward(self, x):
 
 
 # module training with duplicate
-@pytest.mark.parametrize("bs_c", [([1, 2], 64)])
-@pytest.mark.parametrize("same_layer", [True, False])
-def test_module_training_with_duplicate(bs_c, same_layer, device, compile_graph):
+# @pytest.mark.parametrize("bs_c", [([1, 2], 64)])
```
Comment (djl11) [inline ivy_tests/test_ivy/test_stateful/test_modules.py:288]: these comments can be removed

File: ivy_tests/test_ivy/test_stateful/test_modules.py [hunk 7]
```diff
@@ -295,23 +342,31 @@ def __init__(self, in_size, out_size, device=None, hidden_size=64):
...
+        x = ivy.expand_dims(x, axis=0)
         x = ivy.tanh(self._layers["linear0"](x))
         x = ivy.tanh(self._layers["linear1"](x))
         return ivy.tanh(self._layers["linear2"](x))[0]
 
 
 # module with dict training
-@pytest.mark.parametrize("bs_ic_oc", [([1, 2], 4, 5)])
-def test_module_w_dict_training(bs_ic_oc, device, compile_graph):
+# @pytest.mark.parametrize("bs_ic_oc", [([1, 2], 4, 5)])
```
Comment (djl11) [inline ivy_tests/test_ivy/test_stateful/test_modules.py:355]: this comments can be removed

File: ivy_tests/test_ivy/test_stateful/test_modules.py [hunk 8]
```diff
@@ -295,23 +342,31 @@ def __init__(self, in_size, out_size, device=None, hidden_size=64):
...
+    output_channels=st.integers(min_value=2, max_value=5),
+)
+def test_module_w_dict_training(batch_shape, input_channels, output_channels, device):
     # smoke test
     if ivy.current_backend_str() == "numpy":
         # NumPy does not support gradients
-        pytest.skip()
-    batch_shape, input_channels, output_channels = bs_ic_oc
+        return
+    # batch_shape, input_channels, output_channels = bs_ic_oc
```
Comment (djl11) [inline ivy_tests/test_ivy/test_stateful/test_modules.py:369]: this comments can be removed

File: ivy_tests/test_ivy/test_stateful/test_modules.py [hunk 9]
```diff
@@ -355,23 +410,33 @@ def __init__(self, in_size, out_size, device=None, hidden_size=64):
...
         return ivy.Container(x=self._linear0.v, y=self._linear1.v, z=self._linear2.v)
 
     def _forward(self, x):
         pass
 
 
 # with custom var structure
-@pytest.mark.parametrize("bs_ic_oc", [([1, 2], 4, 5)])
-def test_with_custom_var_structure(bs_ic_oc, device):
+# @pytest.mark.parametrize("bs_ic_oc", [([1, 2], 4, 5)])
```
Comment (djl11) [inline ivy_tests/test_ivy/test_stateful/test_modules.py:423]: this comments can be removed

File: ivy_tests/test_ivy/test_stateful/test_modules.py [hunk 10]
```diff
@@ -355,23 +410,33 @@ def __init__(self, in_size, out_size, device=None, hidden_size=64):
...
+def test_with_custom_var_structure(
+    batch_shape, input_channels, output_channels, device
+):
     # smoke test
     if ivy.current_backend_str() == "numpy":
         # NumPy does not support gradients
-        pytest.skip()
-    batch_shape, input_channels, output_channels = bs_ic_oc
+        return
+    # batch_shape, input_channels, output_channels = bs_ic_oc
```
Comment (djl11) [inline ivy_tests/test_ivy/test_stateful/test_modules.py:439]: this comments can be removed

File: ivy_tests/test_ivy/test_stateful/test_modules.py [hunk 11]
```diff
@@ -404,13 +469,21 @@ def _forward(self, x):
 
 
 # top variables
-@pytest.mark.parametrize("bs_ic_oc", [([1, 2], 4, 5)])
-def test_top_variables(bs_ic_oc, device):
+# @pytest.mark.parametrize("bs_ic_oc", [([1, 2], 4, 5)])
```
Comment (djl11) [inline ivy_tests/test_ivy/test_stateful/test_modules.py:472]: this comments can be removed

File: ivy_tests/test_ivy/test_stateful/test_modules.py [hunk 12]
```diff
@@ -404,13 +469,21 @@ def _forward(self, x):
...
+    output_channels=st.integers(min_value=2, max_value=5),
+)
+def test_top_variables(batch_shape, input_channels, output_channels, device):
     # smoke test
     if ivy.current_backend_str() == "numpy":
         # NumPy does not support gradients
-        pytest.skip()
-    batch_shape, input_channels, output_channels = bs_ic_oc
+        return
+    # batch_shape, input_channels, output_channels = bs_ic_oc
```
Comment (djl11) [inline ivy_tests/test_ivy/test_stateful/test_modules.py:486]: this comments can be removed

**Review by djl11 — CHANGES_REQUESTED** [review]
Please remove all cases of commented-out old code, otherwise looks good to merge. As you said, the docstrings will require some more work, but that can come later. Also, we should just increase the global default timeout to 7000 I would say.

**Pull request merged**